File: common/src/main/java/org/valkyrienskies/clockwork/AreaData.java
package org.valkyrienskies.clockwork;

import net.minecraft.core.BlockPos;
import org.joml.Vector3ic;
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.SelectedAreaToolkit;

import java.util.Optional;

public interface AreaData {

    static Optional<AreaData> of(Object context) {
        if (context instanceof AreaData) {
            return Optional.of(((AreaData) context));
        }
        return Optional.empty();
    }

    void setArea(SelectedAreaToolkit load);

    SelectedAreaToolkit getArea();

    Optional<BlockPos> getFirstPos();

    void setFirstPos(Optional<BlockPos> pos);

    Optional<BlockPos> getSecondPos();

    void setSecondPos(Optional<BlockPos> pos);

    void shouldReset(boolean reset);
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/util/AreaData.kt
package org.valkyrienskies.clockwork.util

import net.minecraft.core.BlockPos
import org.valkyrienskies.clockwork.content.curiosities.tools.bluper.SelectedAreaToolkit
import java.util.*

interface AreaData {
    companion object {
        @JvmStatic
        fun of(context: Any?): Optional<AreaData> {
            if (context is AreaData) {
                return Optional.of((context as AreaData?)!!)
            }
            return Optional.empty()
        }
    }

    fun setArea(load: SelectedAreaToolkit)

    fun getArea(): SelectedAreaToolkit

    fun getFirstPos(): Optional<BlockPos>

    fun setFirstPos(pos: Optional<BlockPos>)

    fun getSecondPos(): Optional<BlockPos>

    fun setSecondPos(pos: Optional<BlockPos>)

    fun shouldReset(reset: Boolean)

    fun clearAll()
}